core/src/main/scala-3/caliban/schema/ArgBuilderDerivation.scala

```diff
@@ -11,61 +11,79 @@ import scala.deriving.Mirror
+        recurse[names, ts](
+          (
+            constValue[name].toString,
+            Macros.annotations[t],
+            summonInline[ArgBuilder[t]].asInstanceOf[ArgBuilder[Any]]
+          ) :: values
+        )
     }
 
   inline def derived[A]: ArgBuilder[A] =
     inline summonInline[Mirror.Of[A]] match {
       case m: Mirror.SumOf[A] =>
-        lazy val subTypes   = recurse[m.MirroredElemLabels, m.MirroredElemTypes]
+        lazy val subTypes   = recurse[m.MirroredElemLabels, m.MirroredElemTypes]()
         lazy val traitLabel = constValue[m.MirroredLabel]
         new ArgBuilder[A] {
           def build(input: InputValue): Either[ExecutionError, A] =
-            (input match {
-              case EnumValue(value)   => Some(value)
-              case StringValue(value) => Some(value)
-              case _                  => None
-            }) match {
-              case Some(value) =>
-                subTypes.find { (label, annotations, _) =>
-                  annotations.collectFirst { case GQLName(name) => name }.contains(value) || label == value
-                } match {
-                  case Some((_, _, builder)) => builder.asInstanceOf[ArgBuilder[A]].build(InputValue.ObjectValue(Map()))
-                  case None                  => Left(ExecutionError(s"Invalid value $value for trait $traitLabel"))
-                }
-              case None        => Left(ExecutionError(s"Can't build a trait from input $input"))
-            }
+            buildSum[A](subTypes, traitLabel)(input)
         }
 
       case m: Mirror.ProductOf[A] =>
-        lazy val fields      = recurse[m.MirroredElemLabels, m.MirroredElemTypes]
+        lazy val fields      = recurse[m.MirroredElemLabels, m.MirroredElemTypes]()
         lazy val annotations = Macros.paramAnnotations[A].to(Map)
         new ArgBuilder[A] {
           def build(input: InputValue): Either[ExecutionError, A] =
-            fields.map { (label, _, builder) =>
-              input match {
-                case InputValue.ObjectValue(fields) =>
-                  val finalLabel =
-                    annotations.getOrElse(label, Nil).collectFirst { case GQLName(name) => name }.getOrElse(label)
-                  val default    = annotations.getOrElse(label, Nil).collectFirst { case GQLDefault(v) => v }
-                  fields.get(finalLabel).fold(builder.buildMissing(default))(builder.build)
-                case value                          => builder.build(value)
-              }
-            }.foldRight[Either[ExecutionError, Tuple]](Right(EmptyTuple)) { case (item, acc) =>
-              item match {
-                case error: Left[ExecutionError, Any] => error.asInstanceOf[Left[ExecutionError, Tuple]]
-                case Right(value)                     => acc.map(value *: _)
-              }
-            }.map(m.fromProduct)
+            buildProduct(fields, annotations)(input).map(m.fromProduct)
         }
     }
+  end derived
```

Comment:
I'm planning to add the no-indent option, I think this won't be accepted 😅 (yeah, I hate the significant indentation 😆)

Reply:
Removed 😅. On a side note I used to really hate it as well but it's slowly growing on me.